GraphQL schema suggestions

andreas opened the pull request:
As part of reviewing the GraphQL schema, I thought some aspects were slightly confusing as detailed below.

### Nullable fields on `Contents`

The fields `value`, `metadata` and `hash` are optional for the type `Contents`. This reflects that a particular key could be either present or not, but I suggest this should rather be reflected by returning `null` rather than "empty" Contents object.

In practice, this means instead of returning this JSON ...

```json
{
  "someContentsField": {
    "key": "key/not/present",
    "value": null,
    "metadata": null,
    "hash": null
  }
}
```

... return this JSON instead:

```json
{
  "someContentsField": null
}
```

Suggested diff in SDL terms:

```diff
 type Contents {
-  hash: String
+  hash: String!
   key: String!
-  metadata: String
-  value: String
+  metadata: String!
+  value: String!
 }
```

### Relationship between `Node` and `Contents`

The relationship between the types `Node` ("folders") and `Contents` ("files") are not clear to me right now. The type `Node` has fields `value` and `metadata` similar to `Contents`, which makes it "file-like". I think we should either:

#### Option 1: Collapse types

The types `Node` and `Contents` could be collapsed into a single type. This would probably require adding a new field `kind: TreeKind!`, where `TreeKind` is an enum of `NODE` and `CONTENTS`.

Suggested diff in SDL terms:

```diff
-type Contents {
-  ...
-}

-union Tree = Contents | Node

+enum TreeKind {
+  NODE
+  CONTENTS
+}

-type Node {
+type Tree {
  hash: String!
  key: String!
  metadata: String
  tree: [Tree!]!
  value: String
+  kind: TreeKind!
}
```

(the above diff includes renaming `Node` to `Tree`)

This is the simplest approach and somewhat mirrors the types in OCaml, where "files and folders" are not clearly differentiated either IMHO 😛 I don't find it particularly clean though, and I'm concerned it will lead to suboptimal client code.

#### Option 2: More clearly disambiguate

We could more clearly disambiguate between "folders" (`Node`) and "files" (`Contents`). To pursue this option, I think we should remove the fields `value` and `metadata` from `Node`. This requires changing the type of the field `Node.get` to `Tree` rather than `Node` to reflect it could be either a “file or folder”.

Suggested changes in SDL terms:

```diff
 type Node {
-  get(step: Step): Node
+  get(key: Key!): Tree
   hash: String!
   key: String!
-  metadata: String
   tree: [Tree!]!
-  value: String
 }
```

If we pursue this approach, I think we should also consider how to make the following fields and their types more consistent:

```graphql
type Branch {
  get(key: Key!): String
  get_tree(key: Key!): [Contents!]
  get_all(key: Key!): Contents
  ...
}

type Node {
  get(key: Key!): Tree
  ...
}
```

### Changes in this PR

The changes in this PR include:

- Making fields on `Contents` non-nullable.
- Implementing option 2 (disambiguation) by removing fields `value` and `metadata` from `Node` and changing the type of the field `Node.get`.
- Getting rid of the scalar type `Step`, which was only used in a single instance, which I found inconsistent.
- The source type of Node is not changed (`tree * key`), but the semantics change: the tree is now fully resolved to the actual subtree, and the key is only part of the source type to support the `key` field.

zshipko commented on src/irmin-graphql/server.ml:
```diff
@@ -227,58 +222,46 @@ module Make_ext(Server: Cohttp_lwt.S.Server)(Config: CONFIG)(Store : Irmin.S)(Pr
+                          Store.Tree.get_all tree key' >|= fun (c, m) ->
+                          Lazy.(force contents_as_tree (c, m, key'))
+                        | `Node ->
+                          Store.Tree.get_tree tree key >|= fun t ->
```
During testing this line seemed to be raising error when it shouldn't. I tried replacing `key` with `(Store.Key.v [step])` and that gave me the results I was expecting, but maybe that was just coincidental based on my single example.

andreas replied:
Ah thanks, it's supposed `key'` not `key`.

andreas commented:
@zshipko, I fixed the issue you pointed out and added another suggestion:

```diff
type query {
-  branches: [String!]!
+  branches: [Branch!]!
  ...
}
```

What do you think of that?

zshipko commented:
That seems reasonable to me. I will do one more round of testing then merge this if everything looks good!
